### PR TITLE
Add end-to-end test for max init code size limit and meter

### PR DESCRIPTION
**Pull request by facuMH**
This PR adds an end-to-end test that stress the charging and limits of init code size ([EIP-3860](https://eips.ethereum.org/EIPS/eip-3860))

The test added performs the following steps:

- deploys a smart contract capable attempts to create a new contract with the provided init code size, and reports how much this costed via a log message. 

- creates 3 contracts that initialize to 1, 2 and 3 32 byte words, and checks that the difference in costs is as expected.

- attempts to create a contract where the available gas is not enough to cover for the init code size.

- creates a contract using the max size for init code.

- attempts to create a contract using a code size larger than allowed with enough gas to pay for it. 

This PR is part of https://github.com/Fantom-foundation/sonic-admin/issues/21

### CODE REVIEW

File: tests/initcoder_test.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	r.NoError(err)
+	return log.Cost.Uint64(), receipt.GasUsed
+}
+
+func createContractWithCodeLenAndGas(r *require.Assertions, net *IntegrationTestNet, contract *initcode.Initcode, codeLen, gasLimit uint64) *types.Receipt {
+	receipt, err := net.Apply(func(opts *bind.TransactOpts) (*types.Transaction, error) {
+		opts.GasLimit = gasLimit
+		return contract.CreateContractWith(opts, big.NewInt(int64(codeLen)))
+	})
+	fmt.Printf("gas provided: %v, and gas used: %v\n", gasLimit, receipt.GasUsed) // THIS IS A DEBUG LINE
```
Comment (HerbertJordan) [inline tests/initcoder_test.go:75]: nit: remove

File: tests/initcoder_test.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	// Deploy the invalid start contract.
+	contract, receipt, err := DeployContract(net, initcode.DeployInitcode)
+	r.NoError(err)
+	r.Equal(types.ReceiptStatusSuccessful, receipt.Status, "failed to deploy contract")
+
+	const gasForOneWord uint64 = 54977
+
+	t.Run("create transaction with enough gas for init code", func(t *testing.T) {
+		// we need to provide same gas to all transactions so that we can compare the cost.
+		cost1Word, _ := createAndCheckContractWithCodeLenAndGas(r, net, contract, 30, gasForOneWord+4)
```
Comment (HerbertJordan) [inline tests/initcoder_test.go:30]: nit: can you include the case for 0 words? This would be an interesting edge case as well.

Comment (HerbertJordan) [inline tests/initcoder_test.go:30]: for documentation, it would be good to define a constant
```
const gasPerInitWord = 2
```
and use this constant here and below to document where the constants 2 and 4 are coming from.

File: tests/initcoder_test.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+	})
+
+	t.Run("create transaction without enough gas for init code", func(t *testing.T) {
+		receipt = createContractWithCodeLenAndGas(r, net, contract, 42, gasForOneWord)
+		r.Equal(types.ReceiptStatusFailed, receipt.Status,
+			"unexpectedly succeeded to create contract without enough gas")
+	})
+
+	const MAX_INITCODE_SIZE = 49152
+	const MAX_INITCODE_COST = (MAX_INITCODE_SIZE / 32) * 2
```
Comment (HerbertJordan) [inline tests/initcoder_test.go:48]: also here, 2 should be replaced by a constant

File: tests/initcoder_test.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	const someCost = gasForOneWord + MAX_INITCODE_COST + 9_199
```
Comment (HerbertJordan) [inline tests/initcoder_test.go:49]: If I understand correctly, this should not be `someCost` but `sufficientGas`

**Review by HerbertJordan — CHANGES_REQUESTED** [review]
The test seems to miss two additional ways of creating contracts:
 - creation using `create2` (likely to use the same code path as `create`, but in a black-box test like this one should not assume that)
 - creation of a contract using a transaction without a target; I assume also this should be limited and metered.

File: tests/contracts/initcode/initcode.sol [hunk 5]
```diff
@@ -0,0 +1,14 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.8;
+
+contract InitCode {
+    event LogCost(uint256 cost);
+    
+    function createContractWith(uint codeSize) public {
+        uint256 before = gasleft();
+        assembly ("memory-safe") {
+            pop(create(0, 0, codeSize))
```
Comment (LuisPH3) [inline tests/contracts/initcode/initcode.sol:10]: This code is too obscure, can you document what is going on?

File: tests/initcoder_test.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+	net, err := StartIntegrationTestNet(t.TempDir())
+	r.NoError(err)
+	defer net.Stop()
+
+	// Deploy the invalid start contract.
+	contract, receipt, err := DeployContract(net, initcode.DeployInitcode)
+	r.NoError(err)
+	r.Equal(types.ReceiptStatusSuccessful, receipt.Status, "failed to deploy contract")
+
+	const gasForOneWord uint64 = 54977
```
Comment (LuisPH3) [inline tests/initcoder_test.go:26]: where does this come from?

File: tests/initcoder_test.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+	t.Run("create transaction with enough gas for init code", func(t *testing.T) {
+		// we need to provide same gas to all transactions so that we can compare the cost.
+		cost1Word, _ := createAndCheckContractWithCodeLenAndGas(r, net, contract, 30, gasForOneWord+4)
+		cost2Words, _ := createAndCheckContractWithCodeLenAndGas(r, net, contract, 42, gasForOneWord+4)
+		cost3Words, _ := createAndCheckContractWithCodeLenAndGas(r, net, contract, 90, gasForOneWord+4)
```
Comment (LuisPH3) [inline tests/initcoder_test.go:32]: this test should be a table

File: tests/initcoder_test.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+		}
+	})
+
+	t.Run("create transaction without enough gas for init code", func(t *testing.T) {
+		receipt = createContractWithCodeLenAndGas(r, net, contract, 42, gasForOneWord)
+		r.Equal(types.ReceiptStatusFailed, receipt.Status,
+			"unexpectedly succeeded to create contract without enough gas")
+	})
+
+	const MAX_INITCODE_SIZE = 49152
```
Comment (LuisPH3) [inline tests/initcoder_test.go:47]: mixing capiltalized and not capitalized symbols. Previous constant used other naming convention. Please use camelCase.

File: tests/initcoder_test.go (hunk 4 above)
Comment (LuisPH3) [reply tests/initcoder_test.go:49]: where does 9199 come from?

File: tests/initcoder_test.go
Comment (LuisPH3) [inline tests/initcoder_test.go:1]: This file introduces a pattern :
```
call_test(delcaration)
call_test(delcaration)
call_test(delcaration)

 func call_test(..)  // test logic
```

But we write tests like this:
```
table = [
declaration
declaration
declaration
]

for test in table{
     // test logic
}
```

File: tests/initcoder_test.go [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+func createAndCheckContractWithCodeLenAndGas(r *require.Assertions, net *IntegrationTestNet, contract *initcode.Initcode, codeLen, gasLimit uint64) (uint64, uint64) {
+	receipt := createContractWithCodeLenAndGas(r, net, contract, codeLen, gasLimit)
+	r.Equal(types.ReceiptStatusSuccessful, receipt.Status, "failed to create contract with code length %d", codeLen)
+
+	log, err := contract.ParseLogCost(*receipt.Logs[0])
+	r.NoError(err)
+	return log.Cost.Uint64(), receipt.GasUsed
+}
+
+func createContractWithCodeLenAndGas(r *require.Assertions, net *IntegrationTestNet, contract *initcode.Initcode, codeLen, gasLimit uint64) *types.Receipt {
```
Comment (LuisPH3) [inline tests/initcoder_test.go:70]: nit: the variable should always be called require, not r.

File: tests/initcoder_test.go [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+	"testing"
+
+	"github.com/Fantom-foundation/go-opera/tests/contracts/initcode"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/stretchr/testify/require"
+)
+
+func TestInitCodeSizeLimitAndMetered(t *testing.T) {
+	r := require.New(t)
```
Comment (LuisPH3) [inline tests/initcoder_test.go:15]: nit: the variable should be called require, not r

**Review by LuisPH3 — CHANGES_REQUESTED** [review]
I think this test is weak, it depends too much on the code generation, which could easily change the static costs of the hard-coded variables.
An approach where different executions compare to each other could factor out implementation costs, leaving the difference between executions be the gas measurement wanted.

File: tests/initcoder_test.go (hunk 7 above)
Comment (facuMH) [reply tests/initcoder_test.go:32]: I don't think so, because I need to compare the result of each call, I'm not only check that the calls are successful, but also cross checking between results.

File: tests/initcoder_test.go (hunk 8 above)
Comment (facuMH) [reply tests/initcoder_test.go:47]: the idea here was to differentiate between constants taken from EIP, and constants that I made for the test.

File: tests/initcoder_test.go (hunk 6 above)
Comment (facuMH) [reply tests/initcoder_test.go]: I'm changing the test to better show where this value comes from.

File: tests/initcoder_test.go
Comment (facuMH) [reply tests/initcoder_test.go:1]: I don't think table test style can be applied here because, for `"create transaction with enough gas for init code"` I want to compare each result, not just check that the deploy has been successful, while for the other 3 cases, 2 are fail check, and 1 is a regular success check.

File: tests/initcoder_test.go (hunk 4 above)
Comment (facuMH) [reply tests/initcoder_test.go]: this number was part of the initialization cost of the init code size. it was been rework so that the numbers and intents are clearer.

File: tests/contracts/initcode/initcode.sol [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+        emit LogCost(before - gasleft());
+    }
+
+    function create2ContractWith(uint codeSize) public {
+        // save gas before attempting to create
+        uint256 before = gasleft();
+        uint256 result = 0;
+        assembly  {
+            // top level expresssions are not supposed to return a value, but create does.
+            // so we need to call pop to catch this value.
```
Comment (LuisPH3) [inline tests/contracts/initcode/initcode.sol:25]: nit: we need to assign the value

**Review by LuisPH3 — APPROVED** [review]

**Review by LuisPH3 — DISMISSED** [review]
:+1:

File: tests/initcoder_test.go [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+		opts.GasLimit = gasLimit
+		return variant(opts, big.NewInt(int64(codeLen))) //
+	})
+	require.NoError(err)
+	return receipt
+}
+
+type variant func(opts *bind.TransactOpts, codeSize *big.Int) (*types.Transaction, error)
+
+func runTransactionWithCodeSizeAndGas(r *require.Assertions, net *IntegrationTestNet, codeSize, gas uint64) (*types.Receipt, error) {
```
Comment (LuisPH3) [inline tests/initcoder_test.go:130]: nit:
r => require

File: tests/initcoder_test.go [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+	// these values are needed for the transaction but are irrelevant for the test
+	client, err := net.GetClient()
+	r.NoError(err, "failed to connect to the network:")
```
Comment (LuisPH3) [inline tests/initcoder_test.go:133]: The connection to the client should be closed. 
Just like other uses inside the integration net API.

defer client.Close()

Comment (facuMH) [reply tests/initcoder_test.go:133]: I closed it manually in line 144, because the call to `net.Run` in line 156 internally makes another call to `GetClient` and this sometimes resulted in a timeout. closing the client before calling run, stopped the timeouts. But I have to remove the commented defer, this was an oversight, sorry.

File: tests/initcoder_test.go [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+		receipt2, err := runTransactionWithCodeSizeAndGas(require, net, 2, sufficientGas)
+		require.NoError(err)
+		require.Equal(types.ReceiptStatusSuccessful, receipt2.Status, "failed on transfer to empty receiver with valid code")
+
+		difference := receipt2.GasUsed - receipt1.GasUsed
+		require.Equal(difference, zeroByteCost, "gas difference between 1 and 2 words should be 4, instead got %d",
+			difference)
+	})
+
+	t.Run("make a transaction without enough gas for init code size", func(t *testing.T) {
```
Comment (LuisPH3) [inline tests/initcoder_test.go:67]: I don't understand this one. 
```
	transaction, err := types.SignTx(types.NewTx(&types.AccessListTx{
		ChainID:  chainId,
		Gas:      gas,
		GasPrice: price,
		To:       nil,
		Nonce:    nonce,
		Data:     make([]byte, codeSize),
```

This creates a transaction without code. and allocates a buffer for the data (input for the contract to be executed). It makes sense that it fails because of not having enough gas, but I do not understand what does this have to do with CREATE/CREATE2 
Can you develop?

File: tests/initcoder_test.go [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+	require.NoError(err)
+	require.Equal(types.ReceiptStatusSuccessful, receipt.Status, "failed to deploy contract")
+
+	// deploy measureGasAndAssign to get cost of deploying a contract without create.
+	assignCost, gasUsed := createAndCheckContractWithCodeLenAndGas(require, net, contract, contract.MeasureGasAndAssign, 0, 0_000)
+
+	// -- using CREATE instruction
+	const wordCostCreate uint64 = 2
+	var gasForCreate uint64 = 32000 + gasUsed + assignCost
+	t.Run("create", func(t *testing.T) {
```
Comment (LuisPH3) [inline tests/initcoder_test.go:36]: because of variable name  - package name clash, it would be better to create the require variable inside of the function.

File: tests/initcoder_test.go [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+		opts.GasLimit = gasLimit
+		return variant(opts, big.NewInt(int64(codeLen))) //
+	})
+	require.NoError(err)
+	return receipt
+}
+
+type variant func(opts *bind.TransactOpts, codeSize *big.Int) (*types.Transaction, error)
+
+func runTransactionWithCodeSizeAndGas(require *require.Assertions, net *IntegrationTestNet, codeSize, gas uint64) (*types.Receipt, error) {
```
Comment (LuisPH3) [inline tests/initcoder_test.go:130]: pass t and create require inside for better error location reporting.

File: tests/initcoder_test.go [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+
+func createAndCheckContractWithCodeLenAndGas(require *require.Assertions, net *IntegrationTestNet, contract *initcode.Initcode, variant variant, codeLen, gasLimit uint64) (uint64, uint64) {
+	receipt := createContractWithCodeLenAndGas(require, net, variant, codeLen, gasLimit)
+	require.Equal(types.ReceiptStatusSuccessful, receipt.Status, "failed to create contract with code length %d", codeLen)
+	log, err := contract.ParseLogCost(*receipt.Logs[0])
+	require.NoError(err)
+	return log.Cost.Uint64(), receipt.GasUsed
+}
+
+func createContractWithCodeLenAndGas(require *require.Assertions, net *IntegrationTestNet, variant variant, codeLen, gasLimit uint64) *types.Receipt {
```
Comment (LuisPH3) [inline tests/initcoder_test.go:119]: pass t and create require inside for better error location reporting.

File: tests/initcoder_test.go [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+	})
+
+	t.Run("create transaction with MAX_INITCODE_SIZE+1", func(t *testing.T) {
+		receipt = createContractWithCodeLenAndGas(require, net, variant, MAX_INIT_CODE_SIZE+1, sufficientGas)
+		require.Equal(types.ReceiptStatusFailed, receipt.Status,
+			"unexpectedly succeeded to create contract with init code length greater than MAX_INITCODE_SIZE")
+	})
+}
+
+func createAndCheckContractWithCodeLenAndGas(require *require.Assertions, net *IntegrationTestNet, contract *initcode.Initcode, variant variant, codeLen, gasLimit uint64) (uint64, uint64) {
```
Comment (LuisPH3) [inline tests/initcoder_test.go:111]: pass t and create require inside for better error location reporting.

**Review by LuisPH3 — CHANGES_REQUESTED** [review]
Unfortunatelly I do not understand the last 2 cases, I think they are not related to CREATE, and we may be testing something else.
Please take a look on them because if there is a property from CREATE that we should be testing, I do not think we manage to stimulate it via the transactions created.

File: tests/initcoder_test.go [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+	})
+
+	t.Run("create transaction without enough gas for init code", func(t *testing.T) {
+		receipt = createContractWithCodeLenAndGas(require, net, variant, 30, gasForContract-wordCost)
+		require.Equal(types.ReceiptStatusFailed, receipt.Status,
+			"unexpectedly succeeded to create contract without enough gas")
+	})
+
+	t.Run("create transaction with max init code size", func(t *testing.T) {
+		_, _ = createAndCheckContractWithCodeLenAndGas(require, net, contract, variant, MAX_INIT_CODE_SIZE, sufficientGas)
```
[the remark this answers is not in the document]
Comment (facuMH) [reply tests/initcoder_test.go:101]: both returns are numerical values that are unrelated for this tests case. in this case we just want to check that it can run with those parameters and no error occurred. Errors are checked inside the function.

File: tests/initcoder_test.go (hunk 14 above)
Comment (facuMH) [reply tests/initcoder_test.go:67]: this case is, as you mention, unrealted to CREATE/CREATE2. it is only testing that in a transaction where the init code is less than `MAX_INIT_CODE_SIZE`, it is charged for the of it.

Comment (LuisPH3) [reply tests/initcoder_test.go:67]: I get it now, the mentioned "create transaction". thanks it makes more sense now.

File: tests/initcoder_test.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (LuisPH3) [reply tests/initcoder_test.go:101]: errcheck will only report non-catched errors, other ignored return values can be ignored.

File: tests/initcoder_test.go (hunk 14 above)
Comment (LuisPH3) [reply tests/initcoder_test.go:67]: as discussed offline, it is clear what this does not.

File: tests/initcoder_test.go [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+	// -- using CREATE2 instruction
+	// create2 has extra costs for hashing, this is explained in
+	// https://eips.ethereum.org/EIPS/eip-3860 and reflected in evm.codes calculator.
+	const wordCostCreate2 uint64 = wordCostCreate + 6
+	var gasForCreate2 uint64 = gasForCreate + 44
+	t.Run("create2", func(t *testing.T) {
+		testForVariant(t, net, contract, receipt, contract.Create2ContractWith, gasForCreate2, wordCostCreate2)
+	})
+
+	t.Run("make a transaction that charges for the init code size", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (LuisPH3) [reply tests/initcoder_test.go:49]: i see it now, these transactions create, follow semantics of the create operations, but they do not execute any create opcode

File: tests/initcoder_test.go [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+		cost1Word, _ := createAndCheckContractWithCodeLenAndGas(require, net, contract, variant, 30, gasFor3Words)
+		cost2Words, _ := createAndCheckContractWithCodeLenAndGas(require, net, contract, variant, 42, gasFor3Words)
+		cost3Words, _ := createAndCheckContractWithCodeLenAndGas(require, net, contract, variant, 90, gasFor3Words)
+		require.Equal(cost2Words-cost1Word, wordCost,
+			"cost difference between 1 and 2 words should be %v, instead got %d", wordCost, cost2Words-cost1Word)
+		require.Equal(cost3Words-cost1Word, gasCostFor2Words,
+			"cost difference between 1 and 3 words should be %v, instead got %d", gasCostFor2Words, cost3Words-cost1Word)
+	})
+
+	t.Run("create transaction without enough gas for init code", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (facuMH) [reply tests/initcoder_test.go:94]: the property I want to check here is not whether an account is created or not, but rather that "transaction fails without  enough gas". I will rename to that if you think is appropiate.

File: tests/initcoder_test.go [hunk 22]
```diff
@@ -0,0 +1,170 @@
...
+
+	t.Run("aborts with init code size larger than MAX_INITCODE_SIZE", func(t *testing.T) {
+		receipt := createContractWithCodeLenAndGas(t, net, variant, MAX_INIT_CODE_SIZE+1, sufficientGas)
+		require := require.New(t)
+		require.Equal(types.ReceiptStatusFailed, receipt.Status,
+			"unexpectedly succeeded to create contract with init code length greater than MAX_INITCODE_SIZE")
+	})
+}
+
+func createAndCheckContractWithCodeLenAndGas(t *testing.T, net *IntegrationTestNet,
```
Comment (LuisPH3) [inline tests/initcoder_test.go:120]: nit: this function has too many responsibilities, and it feels difficult to read:
 I presume that this function was the result of a refactor that put together repeated code, nevertheless it lacks of a clear goal and responsibility. It just does too many things. 
 
I learned some rule, some long time ago: something like:  You do not do communication with the user inside the computation function, but instead return the result and write all IO (or stdout for the time being) at the same layer of the program. 

For this reason it would be better to have some pattern:
```
contract, err := createContract()
checkErr(err)
cost, err := getCost(contract)
checkErr(err)
```
Because in the domain of the unit test, error checking is reporting errors to the user. 
There are some reasons why this is this way, I will try to source some literature.

**Review by LuisPH3 — DISMISSED** [review]

File: tests/initcoder_test.go [hunk 23]
```diff
@@ -0,0 +1,170 @@
...
+	var gasForCreate2 uint64 = gasForCreate + 44
+	t.Run("create2", func(t *testing.T) {
+		testForVariant(t, net, contract, contract.Create2ContractWith, gasForCreate2, wordCostCreate2)
+	})
+
+	t.Run("create transaction", func(t *testing.T) {
+		t.Run("charges depending on the init code size", func(t *testing.T) {
+			// transactions charge 4 gas for each zero byte in data.
+			const zeroByteCost uint64 = 4
+			// create a transaction with 1 word of code.
```
Comment (HerbertJordan) [inline tests/initcoder_test.go:53]: It looks like the comments and error messages in this section are mixing up words and bytes ... please align them.

File: tests/initcoder_test.go [hunk 24]
```diff
@@ -0,0 +1,170 @@
...
+			require.Equal(difference, zeroByteCost,
+				"gas difference between 1 and 2 words should be 4, instead got", difference)
+		})
+
+		t.Run("aborts with init code size larger than MAX_INITCODE_SIZE", func(t *testing.T) {
+			// as specified in https://eips.ethereum.org/EIPS/eip-3860#rules,
+			// this is similar to transactions considered invalid for not meeting the intrinsic gas cost requirement.
+			_, err := runTransactionWithCodeSizeAndGas(t, net, MAX_INIT_CODE_SIZE+1, sufficientGas)
+			require.ErrorContains(err,
+				"intrinsic gas too low", "unexpectedly succeeded to create contract with init code length greater than MAX_INITCODE_SIZE")
```
Comment (HerbertJordan) [inline tests/initcoder_test.go:75]: If the error is "intrinsic gas too low" then the transaction was not stopped by the code-size limit but by insufficient gas. I do not think that both of these errors should be accepted here. Only the second.

Is `require.ErrorContains` a either or or a all-of-those condition?

File: tests/initcoder_test.go [hunk 25]
```diff
@@ -0,0 +1,170 @@
...
+	})
+
+	t.Run("create transaction", func(t *testing.T) {
+		t.Run("charges depending on the init code size", func(t *testing.T) {
+			// transactions charge 4 gas for each zero byte in data.
+			const zeroByteCost uint64 = 4
+			// create a transaction with 1 word of code.
+			receipt1, err := runTransactionWithCodeSizeAndGas(t, net, 1, sufficientGas)
+			require.NoError(err)
+			require.Equal(types.ReceiptStatusSuccessful, receipt1.Status,
```
Comment (HerbertJordan) [inline tests/initcoder_test.go:56]: If I understand the `require` library correctly, you should re-create a new `require` instance in every sub-test because otherwise `require` would report issues to the wrong `t *testing.T`.

File: tests/initcoder_test.go [hunk 26]
```diff
@@ -0,0 +1,170 @@
...
+	const wordCostCreate2 uint64 = wordCostCreate + 6
+	var gasForCreate2 uint64 = gasForCreate + 44
+	t.Run("create2", func(t *testing.T) {
+		testForVariant(t, net, contract, contract.Create2ContractWith, gasForCreate2, wordCostCreate2)
+	})
+
+	t.Run("create transaction", func(t *testing.T) {
+		t.Run("charges depending on the init code size", func(t *testing.T) {
+			// transactions charge 4 gas for each zero byte in data.
+			const zeroByteCost uint64 = 4
```
Comment (HerbertJordan) [inline tests/initcoder_test.go:52]: It would be great to also check that non-zero values are correctly charged.

**Review by HerbertJordan — CHANGES_REQUESTED** [review]

File: tests/initcoder_test.go [hunk 27]
```diff
@@ -0,0 +1,170 @@
...
+const sufficientGas = uint64(100_000)
+
+func TestInitCodeSizeLimitAndMetered(t *testing.T) {
+	require := require.New(t)
+
+	net, err := StartIntegrationTestNet(t.TempDir())
+	require.NoError(err)
+	defer net.Stop()
+
+	// Deploy the invalid start contract.
```
[the remark this answers is not in the document]
Comment (facuMH) [reply tests/initcoder_test.go:25]: leftover comment from previous implementations. will be deleted.

File: tests/contracts/initcode/initcode.sol [hunk 28]
```diff
@@ -0,0 +1,42 @@
...
+        emit LogCost(before - gasleft());
+    }
+
+    function Create2ContractWith(uint codeSize) public {
+        // save gas before attempting to create
+        uint256 before = gasleft();
+        uint256 result = 0;
+        assembly {
+            // top level expresssions are not supposed to return a value, but create does.
+            // so we need to assign its return value to catch it.
```
[the remark this answers is not in the document]
Comment (facuMH) [reply tests/contracts/initcode/initcode.sol:26]: I will change this comment to explain in a simpler manner why we need to store the result of create.

File: tests/initcoder_test.go (hunk 26 above)
Comment (facuMH) [reply tests/initcoder_test.go:52]: to make this test, it would probably be best to make a different solidity contract. This might come later.

File: tests/initcoder_test.go (hunk 24 above)
Comment (facuMH) [reply tests/initcoder_test.go:75]: as the comment above says, the first 2 rules in the EIP say that intrinsic gas error is the expected one. Or at least that was my interpretation. 

` ErrorContains(theError error, contains string, msgAndArgs ...interface{})` is the signature, I think this answers your quesiton.

File: tests/initcoder_test.go (hunk 22 above)
Comment (facuMH) [reply tests/initcoder_test.go]: I hope the new changes move towards what you comment here

File: tests/initcoder_test.go [hunk 29]
```diff
@@ -0,0 +1,184 @@
...
+	require.Equal(types.ReceiptStatusSuccessful, receipt.Status, "failed to create contract with code length ", codeLen)
+	return receipt
+}
+
+func createContractWithCodeLenAndGas(t *testing.T, net *IntegrationTestNet, variant variant, codeLen, gasLimit uint64) *types.Receipt {
+	receipt, err := net.Apply(func(opts *bind.TransactOpts) (*types.Transaction, error) {
+		opts.GasLimit = gasLimit
+		return variant(opts, big.NewInt(int64(codeLen)))
+	})
+	if err != nil {
```
Comment (LuisPH3) [inline tests/initcoder_test.go:141]: nit: mixing require and testing apis

File: tests/initcoder_test.go [hunk 30]
```diff
@@ -0,0 +1,184 @@
...
+		t.Errorf("failed to create contract: %v", err)
+	}
+	return receipt
+}
+
+func getCreateCost(t *testing.T, contract *contractcreator.Contractcreator, receipt *types.Receipt) uint64 {
+	log, err := contract.ParseLogCost(*receipt.Logs[0])
+	if err != nil {
```
Comment (LuisPH3) [inline tests/initcoder_test.go:149]: nit:  mixing require and testing apis

File: tests/initcoder_test.go [hunk 31]
```diff
@@ -0,0 +1,184 @@
...
+	contract, receipt, err := DeployContract(net, contractcreator.DeployContractcreator)
+	requireBase.NoError(err)
+	requireBase.Equal(types.ReceiptStatusSuccessful, receipt.Status, "failed to deploy contract")
+
+	// deploy measureGasAndAssign to get cost of deploying a contract without create.
+	receipt = createContractSuccessfully(t, net, contract.GetOverheadCost, 0, 0)
+
+	// -- using CREATE instruction
+	const wordCostCreate uint64 = 2
+	var gasForCreate uint64 = 32035 + receipt.GasUsed //+ assignCost
```
Comment (LuisPH3) [inline tests/initcoder_test.go:35]: nit: this comment is not clear, is this already added, or is it a comment to the formula?

File: tests/initcoder_test.go [hunk 32]
```diff
@@ -0,0 +1,184 @@
...
+	// create2 has extra costs for hashing, this is explained in
+	// https://eips.ethereum.org/EIPS/eip-3860 and reflected in evm.codes calculator.
+	const wordCostCreate2 uint64 = wordCostCreate + 6
+	var gasForCreate2 uint64 = gasForCreate
+	t.Run("create2", func(t *testing.T) {
+		testForVariant(t, net, contract, contract.Create2With, gasForCreate2, wordCostCreate2)
+	})
+
+	t.Run("create transaction", func(t *testing.T) {
+		t.Run("charges depending on the init code size", func(t *testing.T) {
```
Comment (LuisPH3) [inline tests/initcoder_test.go:50]: nit: this case could be a function, like the others.
That would solve the alias problem with requireBase and the nested require variables.

File: tests/initcoder_test.go (hunk 29 above)
Comment (facuMH) [reply tests/initcoder_test.go:141]: did not think it was worth to instantiate a new `require` just for this one simple check.

**Review by jenikd — CHANGES_REQUESTED** [review]

File: tests/initcoder_test.go [hunk 33]
```diff
@@ -0,0 +1,191 @@
...
+	var gasForCreate uint64 = 32035 + receipt.GasUsed
+	t.Run("create", func(t *testing.T) {
+		testForVariant(t, net, contract, contract.CreatetWith, gasForCreate, wordCostCreate)
+	})
+
+	// -- using CREATE2 instruction
+	// create2 has extra costs for hashing, this is explained in
+	// https://eips.ethereum.org/EIPS/eip-3860 and reflected in evm.codes calculator.
+	const wordCostCreate2 uint64 = wordCostCreate + 6
+	var gasForCreate2 uint64 = gasForCreate
```
[the remark this answers is not in the document]
Comment (facuMH) [reply tests/initcoder_test.go:44]: yes, this is a remnant from some experimentation I was doing, should have been removed.

**Review by jenikd — DISMISSED** [review]
Thanks

File: tests/initcoder_test.go [hunk 34]
```diff
@@ -0,0 +1,191 @@
...
+
+		// createa a transaction with 2 byte of code.
+		receipt2, err := runTransactionWithCodeSizeAndGas(t, net, 2, sufficientGas)
+		require.NoError(err)
+		require.Equal(types.ReceiptStatusSuccessful, receipt2.Status,
+			"failed on transfer to empty receiver with valid code")
+
+		difference := receipt2.GasUsed - receipt1.GasUsed
+		require.Equal(difference, zeroByteCost,
+			"gas difference between 1 and 2 words should be 4, instead got", difference)
```
Comment (HerbertJordan) [inline tests/initcoder_test.go:126]: In the rest of this test you state that code size is measured in bytes, but here it is words. I guess it should be bytes here as well.

File: tests/initcoder_test.go [hunk 35]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+func createContractWithCodeLenAndGas(net *IntegrationTestNet, variant variant, codeLen, gasLimit uint64) (*types.Receipt, error) {
+	return net.Apply(func(opts *bind.TransactOpts) (*types.Transaction, error) {
+		opts.GasLimit = gasLimit
+		return variant(opts, big.NewInt(int64(codeLen)))
+	})
+}
+
+func getCreateCost(contract *contractcreator.Contractcreator, receipt *types.Receipt) (uint64, error) {
```
Comment (HerbertJordan) [inline tests/initcoder_test.go:157]: nit: this is only used once ... inline?

**Review by HerbertJordan — DISMISSED** [review]
Looks good, thanks.

**Review by HerbertJordan — APPROVED** [review]
